### ISSUE TITLE
question: best practices for tests

### ISSUE THREAD

**Issue by alexandruy2k** [gruntwork-io/terratest#684]
Hi,

Apologies for perhaps a silly question as I am new to go and terratest.

I've started to use terratest to remove a lot of the manual tests I used to do for deploying GCP VM instances.
I use packer to build an image of all the software needed, then use terratest to deploy the image and ssh in to ensure that everything is running as expected. This is saving me a lot of time and because the CI doesn't get tired or lazy the same tests will be ran every time.

Because deploying an instance and waiting for it to boot up is taking a good bit of time I'm doing multiple tests over on the same instance and I've been using the code below to split up the tests.

```
var testcases = []struct {
name             string
..........

	for _, testCase := range testcases {
		testCase := testCase

		t.Run(testCase.name, func(t *testing.T) {
			output, err := ssh.CheckSshCommandE(t, host, testCase.command)

	
```

I'm also looking to add RunTestStage so that when developing the tests I can skip destroying the instance so that I get faster feedback.

Is there guidance or best practices for how to decide if tests belong in the same func TestPostDeploymentState(t *testing.T) and when should I create new functions and even when should I create a new test file?

And if I need to split up tests across multiple test functions is there a way to have a GCP instance created globally and shared for all the test cases so that I don't spin up and down the same instance many times and slow down the CI?

I've done a good bit of research so apologies if this is already covered somewhere

**Comment by yorinasub17** [comment]
This is a good question, and indeed we lack some docs on this! Unfortunately, the answer is that there isn't a clear cut recommendation for this, as it ends up being a trade off. Packing multiple tests into a single function has the advantage of reducing cost and speed by reducing duplication, but risk conflating test failures as you may not know what the actual failure is, as most of your tests could be failing for the same reason (e.g., the instance can't call out). It also can be harder to maintain, as you essentially have one giant function.

That said, you can mitigate those concerns to some extent by using test stages and subtests so that you can run a subset of the tests, and also avoid recreating the instance every time, at the expense of increasing the complexity of your test functions.

So it really comes down to what you hope to accomplish with the testing, and what things you are willing to sacrifice/optimize and making the trade offs there.

---
> And if I need to split up tests across multiple test functions is there a way to have a GCP instance created globally and shared for all the test cases so that I don't spin up and down the same instance many times and slow down the CI?

The subtest pattern is how you can achieve this, although there are some gotchas with doing that especially if you are using `t.Parallel`.

Here is a snippet from our (Gruntwork) internal style guide that is relevant to this.



### Setup and Teardown pattern

In some cases you will want to write a group of tests that use a common resource, such as a Docker image or VPC. In this case, you will want to setup the common resource once, run a bunch of tests, and then teardown the resource. To achieve this, you can follow [the subtest pattern](https://blog.golang.org/subtests) of Go.

Always use [table driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) where possible to make the subtest routines maintainable. Briefly, this means that you group your test cases using a test struct that reflects the unique parameters of the test cases. Then you can conveniently loop over the test cases in parallel, taking advantage of uniformity and speed.

Note that the subtest pattern has gotchas when running tests in parallel:

- The main test function will not wait for the subtest to run if it uses `t.Parallel`. To avoid this, you need to wrap the parallel subtests in a synchronous, blocking subtest. In the example below, the `group` subtest is synchronous (no call to `t.Parallel`) and thus the main function will wait until that test finishes. The `group` test does not finish until all the subtests it spawns are finished, even if they are non-blocking and parallel, and thus the `tearDownVPC` call does not happen until all subtests are done.
- If you are using table driven tests, the range variable will be updated to the next iteration before it is used within the subtest. That is, in the example below, if we did not have the `testCase := testCase` line in the range block, the `testCase` reference used in the subtest after the `t.Parallel` call will correspond to the last `testCase` in the `testCases` list. To avoid this, we create a new variable in the scope of the range block so that it does not get
updated during the loop.

Example:

```go
func TestECS(t *testing.T) {
    t.Parallel()

    defer tearDownVPC()
    deployVPC()

    // Wrap the parallel tests in a synchronous test group to ensure that the main test function (the one calling
    // `tearDownVPC` and `deployVPC`) waits until all the subtests are done before running the deferred function.
    t.Run("group", func(t *testing.T) {
        for _, testCase := range testCases {
            // To avoid the range variable from getting updated in the parallel tests, we bind a new name that is within
            // the scope of the for block.
            testCase := testCase
            t.Run(testCase.name, func(t *testing.T) {
                t.Parallel()
                testCase.testCode()
            })
        }
    })
}
```

**Comment by infraredgirl** [comment]
Closing this, as the question has been answered. @alexandruy2k feel free to reopen if needed.
